core/src/main/java/tc/oc/pgm/modules/StatsMatchModule.java

```diff
@@ -0,0 +1,250 @@
+
+      if (hasNoStats(murdererUUID)) murdererStats = putNewPlayer(murdererUUID);
+
+      if (event.getDamageInfo() instanceof ProjectileInfo) {
+        murdererStats.setLongestBowKill(
+            victim
+                .getState()
+                .getLocation()
+                .distance(((ProjectileInfo) event.getDamageInfo()).getOrigin()));
+      }
+
+      murdererStats.onMurder();
+
+      sendLongHotbarMessage(murderer, murdererStats.getBasicStatsMessage());
+    }
+  }
+
+  @EventHandler
+  public void onMatchEnd(MatchFinishEvent event) {
+    Map<UUID, Integer> allKills = new HashMap<>();
+    Map<UUID, Integer> allKillstreaks = new HashMap<>();
+    Map<UUID, Integer> allDeaths = new HashMap<>();
+    Map<UUID, Integer> allBowshots = new HashMap<>();
+
+    for (Map.Entry<UUID, PlayerStats> mapEntry : allPlayerStats.entrySet()) {
+      UUID playerUUID = mapEntry.getKey();
+      PlayerStats playerStats = mapEntry.getValue();
+
+      if (hasNoStats(playerUUID)) playerStats = putNewPlayer(playerUUID);
+
+      allKills.put(playerUUID, playerStats.kills);
+      allKillstreaks.put(playerUUID, playerStats.killstreakMax);
+      allDeaths.put(playerUUID, playerStats.deaths);
+      allBowshots.put(playerUUID, playerStats.longestBowKill);
+    }
+
+    Component killMessage = getMessage("stats.kills", sortStats(allKills), ChatColor.GREEN);
+    Component killstreakMessage =
+        getMessage("stats.killstreak", sortStats(allKillstreaks), ChatColor.GREEN);
+    Component deathMessage = getMessage("stats.deaths", sortStats(allDeaths), ChatColor.RED);
+    TopResult bestBowshot = sortStats(allBowshots);
+    String bowMessageKey = (bestBowshot.stat == 1) ? "stats.bowshot.block" : "stats.bowshot.blocks";
+    Component bowshotMessage = getMessage(bowMessageKey, bestBowshot, ChatColor.YELLOW);
+
+    for (MatchPlayer viewer : match.getPlayers()) {
+      viewer.sendMessage(
+          Components.fromLegacyText(
+              ComponentUtils.horizontalLineHeading(
+                  ChatColor.YELLOW
+                      + AllTranslations.get().translate("stats.best", viewer.getBukkit()),
+                  ChatColor.WHITE,
+                  ComponentUtils.MAX_CHAT_WIDTH)));
+
+      viewer.sendMessage(killMessage);
+      viewer.sendMessage(killstreakMessage);
+      viewer.sendMessage(deathMessage);
+      if (bestBowshot.stat != 0) viewer.sendMessage(bowshotMessage);
+    }
+  }
+
+  private class TopResult {
+    UUID uuid;
+    int stat;
+  }
+
+  private TopResult sortStats(Map<UUID, Integer> map) {
+    Map.Entry<UUID, Integer> mapEntry =
+        map.entrySet().stream().max(Comparator.comparingInt(Map.Entry::getValue)).orElse(null);
+    TopResult topResult = new TopResult();
+
+    if (mapEntry == null) { // Should never happen, but acts as a failsafe
+      topResult.uuid = UUID.fromString("3c7db14d-ac4b-4e35-b2c6-3b2237f382be");
+      topResult.stat = 0;
+      return topResult;
+    }
```

Comment:
You don't need this `null` check, these maps will never be null.

Reply:
Ok!